skxray/cdi.py

```diff
@@ -119,9 +142,284 @@ def convolution(array1, array2):
+        add small value to avoid the case of dividing something by zero
+
+    Returns
+    -------
+    array :
+        updated pattern in q space
+    """
+    diff_tmp = _fft_helper(array_in) / np.sqrt(np.size(array_in))
+    index = np.where(diff_array > 0)
+    diff_tmp[index] = diff_array[index] * diff_tmp[index] / (np.abs(diff_tmp[index]) + thresh_v)
+
+    if pi_modulus_flag == 'Complex':
+        return _fft_helper(diff_tmp, option='ifft') * np.sqrt(np.size(diff_array))
+    elif pi_modulus_flag == 'Real':
+        return np.abs(_fft_helper(diff_tmp, option='ifft')) * np.sqrt(np.size(diff_array))
+
+
+def find_support(sample_obj,
+                 sw_sigma, sw_threshold):
+    """
+    Update sample area based on thresholds.
+
+    Parameters
+    ----------
+    sample_obj : array
+        sample for reconstruction
+    sw_sigma : float
+        sigma for gaussian in shrinkwrap method
+    sw_threshold : float
+        threshold used in shrinkwrap method
+
+    Returns
+    -------
+    new_sup : array
+        updated sample support
+    s_index : array
+        index for sample area
+    s_out_index : array
+        index not for sample area
+    """
+
+    sample_obj = np.abs(sample_obj)
+    gauss_fun = gauss(sample_obj.shape, sw_sigma)
+    gauss_fun = gauss_fun / np.max(gauss_fun)
+
+    conv_fun = convolution(sample_obj, gauss_fun)
+
+    conv_max = np.max(conv_fun)
+
+    s_index = np.where(conv_fun >= (sw_threshold*conv_max))
+    s_out_index = np.where(conv_fun < (sw_threshold*conv_max))
+
+    new_sup = np.zeros_like(sample_obj)
+    new_sup[s_index] = 1
+
+    return new_sup, s_index, s_out_index
+
+
+def pi_support(sample_obj, index_v):
+    """
+    Define sample shape by cutting unnecessary values.
+
+    Parameters
+    ----------
+    sample_obj : array
+        sample data
+    index_v : array
+        index to define sample area
+
+    Returns
+    -------
+    array :
+        sample object with proper cut.
+    """
+    sample_obj = np.array(sample_obj)
+    sample_obj[index_v] = 0
+    return sample_obj
+
+
+def cal_relative_error(x_old, x_new):
+    """
+    Relative error is calculated as the ratio of the difference between the new and
+    the original arrays to the norm of the original array.
+
+    Parameters
+    ----------
+    x_old : array
+        previous data set
+    x_new : array
+        new data set
+
+    Returns
+    -------
+    float :
+        relative error
+    """
+    return np.linalg.norm(x_new - x_old)/np.linalg.norm(x_old)
+
+
+def cal_diff_error(sample_obj, diff_array):
+    """
+    Calculate the error in q space.
+
+    Parameters
+    ----------
+    sample_obj : array
+        sample data
+    diff_array : array
+        diffraction pattern
+
+    Returns
+    -------
+    float :
+        relative error in q space
+    """
+    new_diff = np.abs(_fft_helper(sample_obj)) / np.sqrt(np.size(sample_obj))
+    return cal_relative_error(diff_array, new_diff)
+
+
+class CDI(object):
+
```

Comment:
Doing this as a class makes integrating this with VT very difficult, everything in the library should be functional unless there is a _very_ good reason.